server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -801,10 +819,11 @@ public void onResponse(Releasable releasable) {
                             targetAllocationId
                         );
                         // make sure we release all permits before we resolve the final listener
-                        final ActionListener<Void> wrappedInnerListener = ActionListener.runBefore(
-                            listener,
-                            Releasables.releaseOnce(releasable)::close
-                        );
+                        final ActionListener<Void> wrappedInnerListener = ActionListener.runBefore(listener, () -> {
+                            if (acquiredPrimaryPermits == null) {
```

Comment:
For this `acquiredPrimaryPermits == null` condition and the one below, I wonder if it is necessary now? Maybe we can let the relocated() release the hollow engine's primary permits as well, to reduce code changes here, and deal with it in the context of ES-10573.

Reply:
Yeah, I was thinking about that too. Will revert that change to keep `relocate` implementation simpler.

Reply:
https://github.com/elastic/elasticsearch/pull/120098/commits/f3ca57479206b9f31af3d5880785024f9ca4a8e7